### ISSUE TITLE
[QUESTION] How to upload data to a longhorn persistent volume

### ISSUE THREAD

**Issue by tomsunyu** [longhorn/longhorn#3473]
Hello,
     Besides uploading data to the longhorn persistent volume by executing the command "kubectl copy", is there any other way?
     For example, through NFS, some users can upload data from the nfs server.

**Comment by Thiritin** [comment]
1. You could create a pod with an ftp server image to the volume and access it that way.

2. Try if the "Attach" function works in the dashboard. Check the maintenace box and see if it added a new disk using the lsblk command. (Never worked for me, but perhaps it works on your system.)

**Comment by jenting** [comment]
@tomsunyu 
One question, what's your use case for uploading data to a longhorn persistent volume?
Are you migrating some volume from other storage vendor solutions to longhorn?

**Comment by jenting** [comment]
Probably the backing image upload feature could help. cc @shuo-wu

**Comment by tomsunyu** [comment]
@jenting 
Yes，I am migrating some volume from other storage vendor solutions to longhorn. In addition, some users also need to transfer data to the longhorn volumes. Therefore, I hope to find a simpler and more convenient way.

**Comment by derekbit** [comment]
@Thiritin 
Transfer data from in-cluster pod or from out of the cluster?

**Comment by tomsunyu** [comment]
Recently I encountered another problem. I am going to upgrade kubernetes, but when I execute the following command:
```
kubectl drain k8snode1 --ignore-daemonsets
```
Some longhorn ```instance-r``` ```instance-e``` pods can not be drained. In other words, these pods will start as soon as they are deleted.
So how to solve the problem?

**Comment by tomsunyu** [comment]
> @Thiritin Transfer data from in-cluster pod or from out of the cluster?

yes

**Comment by jenting** [comment]
> Recently I encountered another problem. I am going to upgrade kubernetes, but when I execute the following command:
> 
> ```
> kubectl drain k8snode1 --ignore-daemonsets
> ```
> 
> Some longhorn `instance-r` `instance-e` pods can not be drained. In other words, these pods will start as soon as they are deleted. So how to solve the problem?

Is the replica the last healthy replica? (It means that it's the last replica that contains the data.)

**Comment by tomsunyu** [comment]
> > Recently I encountered another problem. I am going to upgrade kubernetes, but when I execute the following command:
> > ```
> > kubectl drain k8snode1 --ignore-daemonsets
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Some longhorn `instance-r` `instance-e` pods can not be drained. In other words, these pods will start as soon as they are deleted. So how to solve the problem?
> 
> Is the replica the last healthy replica? (It means that it's the last replica that contains the data.)

I redeployed a set of longhorn：

```
NAME                                       READY   STATUS    RESTARTS   AGE    IP              NODE    NOMINATED NODE   READINESS GATES
csi-attacher-5499bf977b-5scfp              1/1     Running   0          10m    172.29.65.17    k8st6   <none>           <none>
csi-provisioner-7c7d964bcb-nrdbp           1/1     Running   0          10m    172.29.99.9     k8st5   <none>           <none>
csi-resizer-796f9597d6-rnbxs               1/1     Running   0          10m    172.29.99.11    k8st5   <none>           <none>
csi-snapshotter-67f87f45d-9rwt8            1/1     Running   0          10m    172.29.65.18    k8st6   <none>           <none>
engine-image-ei-fa8d9882-jmm6n             1/1     Running   0          11m    172.29.99.8     k8st5   <none>           <none>
engine-image-ei-fa8d9882-qdv68             1/1     Running   0          11m    172.29.65.14    k8st6   <none>           <none>
engine-image-ei-fa8d9882-rqbwf             1/1     Running   0          11m    172.29.222.18   k8st4   <none>           <none>
instance-manager-e-6a048e1c                1/1     Running   0          11m    172.29.65.15    k8st6   <none>           <none>
instance-manager-e-8bb1378f                1/1     Running   0          6m6s   172.29.222.21   k8st4   <none>           <none>
instance-manager-e-f99441fc                1/1     Running   0          10m    172.29.99.5     k8st5   <none>           <none>
instance-manager-r-d9a8190c                1/1     Running   0          11m    172.29.65.16    k8st6   <none>           <none>
instance-manager-r-ece1d902                1/1     Running   0          6m     172.29.222.22   k8st4   <none>           <none>
instance-manager-r-f0517c76                1/1     Running   0          10m    172.29.99.6     k8st5   <none>           <none>
longhorn-csi-plugin-4zq9c                  2/2     Running   0          10m    172.29.99.10    k8st5   <none>           <none>
longhorn-csi-plugin-jbskd                  2/2     Running   0          10m    172.29.222.20   k8st4   <none>           <none>
longhorn-csi-plugin-mc8fb                  2/2     Running   0          10m    172.29.65.19    k8st6   <none>           <none>
longhorn-driver-deployer-d4c5887c4-tvqgm   1/1     Running   0          6m9s   172.29.99.12    k8st5   <none>           <none>
longhorn-manager-psx2z                     1/1     Running   0          11m    172.29.65.13    k8st6   <none>           <none>
longhorn-manager-pzjmz                     1/1     Running   0          11m    172.29.99.4     k8st5   <none>           <none>
longhorn-manager-slzwq                     1/1     Running   0          11m    172.29.222.16   k8st4   <none>           <none>
longhorn-ui-75f96446bf-v5r7c               1/1     Running   0          6m9s   172.29.65.20    k8st6   <none>           <none>
```
I mean the instance-manager pods can not be evicted from the previous server, such as instance-manager-e-8bb1378f.

**Comment by innobead** [comment]
> > > Recently I encountered another problem. I am going to upgrade kubernetes, but when I execute the following command:
> > > ```
> > > kubectl drain k8snode1 --ignore-daemonsets
> > >   
> > > Some longhorn `instance-r` `instance-e` pods can not be drained. In other words, these pods will start as soon as they are deleted. So how to solve the problem?
> 
> I mean the instance-manager pods can not be evicted from the previous server, such as instance-manager-e-8bb1378f. When I delete the pod instance-manager-e-8bb1378f, it will be established again. Also, I find no instance-manager daemonsets in the longhorn-system namespace.

This should be related to #3459 cc @shuo-wu

**Comment by jenting** [comment]
> I mean the instance-manager pods can not be evicted from the previous server, such as instance-manager-e-8bb1378f.
When I delete the pod instance-manager-e-8bb1378f, it will be established again. Also, I find no instance-manager daemonsets in the longhorn-system namespace.

Thanks for the reporting. Yes, it's the original design of the instance manager life cycling. And we're working on re-design the instance manager life cycle.
Note that, the instance-manager DaemonSets is managed by longhorn-manager so you can't see it in the `kubectl get ds` output.

**Comment by shuo-wu** [comment]
- Those instance manager pods are simple pods rather than the workloads controlled by a Daemonset.
- If `kubectl drain` succeeds but the engine instance manager pod is still there (re-deployed automatically), this is related to #3459. But once `kubectl drain` gets stuck in evicting the instance manager pods, it means there are running Longhorn volumes on the node and you need to detach them first.

**Comment by tomsunyu** [comment]
> I am going to upgrade kubernete

So do you plan to fix this bug in which version?

**Comment by shuo-wu** [comment]
You mean #3459? The fix will be in v1.3.0 then probably backported to v1.2.4

**Comment by conradkleinespel** [comment]
For reference, this is what I've used to migrate a local volume to a longhorn volume.

First, I delete the pods that used the volume to be migrated.

Then, I create a pod for the purpose of migrating data:
```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: myvolmigpod
  name: myvolmigpod
spec:
  containers:
    - image: busybox
      name: myvolmigpod
      command:
        - sleep
        - "99999999"
      volumeMounts:
        - mountPath: /data/old
          name: old
        - mountPath: /data/new
          name: new
  volumes:
    - name: old
      persistentVolumeClaim:
        claimName: old-pvc
    - name: new
      persistentVolumeClaim:
        claimName: new-pvc
  dnsPolicy: ClusterFirst
  restartPolicy: Always
```

Then, I shell into the pod:
```shell
kubectl exec -ti myvolmigpod /bin/sh
```

Then, I copy files from the old PV to the new PV:
```shell
cp -a /data/old/. /data/new/
```
The `.` after `/data/old/.` is important, see https://superuser.com/a/367303/115129.

Finally, I relaunch the pods with the new volume, verify everything works as expected and delete the old volume.
